fix(challenge-form): fix edit challenge tags (#845)

FlavioKde opened the pull request:
### 🎯 Summary

This PR corresponds to task [#845](https://tree.taiga.io/project/luisvi-ita-challenges/task/845). It is the first task from HU [#816](https://tree.taiga.io/project/luisvi-ita-challenges/us/816), the main goal of this HU is to fix an issue regarding selected tags during challenge edition.

When editing a programming challenge, previously selected tags do not appear active in the form. 
Because of that, if the mentor re-selects the same tag (which is already stored), the frontend sends a duplicated payload such as:

Tags during edition mode:
<img width="886" height="425" alt="image" src="https://github.com/user-attachments/assets/75ee038c-1865-4542-873e-37db3e39a93e" />

Tags selected during edition mode:
<img width="886" height="429" alt="image" src="https://github.com/user-attachments/assets/27e7cb4f-a9b7-43c0-9f22-70bfc675df7c" />



`"tags": ["0c8ca47c-d915-486e-8ae9-803483c8ed17", "0c8ca47c-d915-486e-8ae9-803483c8ed17"]`

This causes a 400 Bad Request from the backend ("tag UUID duplicated: ...") and the user sees no message in the UI.

<img width="886" height="600" alt="image" src="https://github.com/user-attachments/assets/97a8322c-0869-4d2f-a3d3-228f425edfd9" />

--------------------


### 🧠 The fix ensures:

- Stored tags appear preselected (active) when editing a challenge.

- The form sends a deduplicated tag array matching the UI selection.
 
- Validation messages (e.g. “Tags required”) display properly.
 
- The user never sees a silent failure.

--------------------

### 🧩 Key Changes

- Updated `loadTags()` to set `tagsControl` verifying tags initialization

- Updated `onSubmit()` to set `save logic` correctly bringing the persisted tags

- Added unit tests covering both scenarios 

--------------------
### 🔄 Modified files

- /Users/pablofedericogomez/ita-challenges-frontend/src/app/modules/challenge/components/challenge-form/challenge-form.component.ts
 
- /Users/pablofedericogomez/ita-challenges-frontend/src/app/modules/challenge/components/challenge-form/challenge-form.component.spec.ts
 
- /Users/pablofedericogomez/ita-challenges-frontend/package.json

--------------------


### 🔄 Changelog

Changelog updated

--------------------


### 🧪 Final result

- Open the edit form for a challenge with already selected tags during challenge creation→ form loads correctly.

- Open the edit form for a challenge with tags, you will be able to adding or removing tags, and submitting the challenge edited→ endpoint PUT response with a 200 OK.

Edition mode after frontend changes:
<img width="886" height="410" alt="image" src="https://github.com/user-attachments/assets/12afb707-194d-4ffa-80d3-6f79b985dfdc" />


<img width="886" height="429" alt="image" src="https://github.com/user-attachments/assets/99791f54-a5ba-4803-ae18-9144787a3ae6" />

Adding a tag in edition mode after frontend changes:
<img width="886" height="397" alt="image" src="https://github.com/user-attachments/assets/162d37e6-f03d-4d91-8855-ea762b953064" />

API response  after adding changes:
<img width="886" height="552" alt="image" src="https://github.com/user-attachments/assets/efb0a05b-9fc1-459e-8d1d-e035214c9490" />

Outcome:
<img width="886" height="440" alt="image" src="https://github.com/user-attachments/assets/765ccb32-cc1c-4525-90b3-e86b84af8793" />

Removing two tags and adding another one in edition mode after the frontend changes:
<img width="2299" height="1295" alt="image" src="https://github.com/user-attachments/assets/bef2b049-bde6-4be0-b1db-d69678014041" />

Outcome:
<img width="2342" height="1410" alt="image" src="https://github.com/user-attachments/assets/e5721db3-fb1d-45e0-bed5-7a08481b296b" />



- Verify that no console errors appear.


--------------------

###  ✅  PR Author Self-Check

- [x] Tested the flow locally (logged-in vs logged-out).
- [x] Unit tests pass for the new logic.
- [x] No unrelated changes or merge conflicts.

FlavioKde committed:
fix bug:onSubmit, loagTag and their tests/up new version CHANGELOG.md

eduzavarce commented:
From the screenshots, I cannot say whats the original state and what is the outcome.

eduzavarce commented:
From the screenshots, i can assume you can add tags, but i do not see what happens if you revove them.

FlavioKde committed:
docs: add PR number (#709) to CHANGELOG.md

eduzavarce reviewed (commented):
@FlavioKde @Pablo2203 

Everything seems in order as discussed yesterday. 

only one doubt:
<img width="415" height="79" alt="image" src="https://github.com/user-attachments/assets/0665f44b-44bd-4bfe-b518-68d8d391afbf" />

Why isnt this checkbox checked? are there conflicts?

FlavioKde committed:
fix: synchronize tags state between selectedTags and tagsControl

- Add valueChanges subscription to automatically sync selectedTags with tagsControl value
- Remove duplicate state merging in onSubmit method
- Update component tests to reflect synchronized state behavior
- Resolve tag persistence issue when editing challenges

Pablo2203 commented:
@eduzavarce Hello Edu! Thank you for your insights. We agreed with Flavio about those changes you suggested. Now It is everything updated.

- [x] From the screenshots, I cannot say whats the original state and what is the outcome.
**Now there are epigraphs detailing the process states.**

- [x] From the screenshots, i can assume you can add tags, but i do not see what happens if you revove them.
**Examples added showing the tag remotion.**

- [x] Why isnt this checkbox checked? are there conflicts?
**Checkbox checked!**

FlavioKde committed:
test: update tag state synchronization tests

- Verify selectedTags and tagsControl are synchronized in form mapping
- Use tagsControl value directly in createChallenge expectations
- Remove redundant tag merging assertions
- Align tests with synchronized state behavior

FlavioKde committed:
chore: clean up commented code in challenge form

- Remove obsolete Array.from/new Set tag merging code
- Clean up after tag state synchronization implementation

eduzavarce reviewed (approved):
Good job, seems everything is in order now!

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -249,16 +253,19 @@ loadChallengeForEditing(): void {
-    })
-  }
+        
+      this.tagsControl.setValue(this.selectedTags);
```
We could refactor this block to remove duplication and improve error handling:

- Simplify the if/else by computing the array first and calling setValue only once.

FlavioKde replied:
You're absolutely right about the duplication. Currently we have two synchronized states (tagsControl and selectedTags) due to existing technical debt. I kept the synchronization to fix the HU without introducing a larger refactor.

ivilarop replied:
We'll leave that as a technical debt for a future task: "Unify tag selection: single calculation + single setValue (preserve errors/touched/dirty and synchronize selectedTags)." Not blocking here. 

It is included in the previous technical debt

Pablo2203 replied:
@ivilarop Perfect! Let´s do that. Please find the technical debt in this link: https://tree.taiga.io/project/luisvi-ita-challenges/issue/859

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -249,16 +253,19 @@ loadChallengeForEditing(): void {
+            this.tagsControl.setValue([]);
+          }
+        },
+        error: (error) => {
```
- We shouldn’t use throw inside the subscribe error callback.
Since the component already resets its state and gives user feedback, there’s no need to raise an exception here.
Instead, it’s cleaner to handle this earlier in the stream (using catchError) or just log the issue.
This avoids unhandled exceptions and makes the observable flow easier to maintain.

FlavioKde replied:
The throw error is aligned with the rest of the file; it is handled in the same way, so it preserves the semantics of the file.
We can use console.error to align with your solution; it logs the error to the console without breaking the flow.
The component recovers from the error by resetting the state.
If I want to provide feedback to the user, I should modify the HTML,
which would be outside the scope of our responsibility in this user story.

In the onSubmit() method, I use commonModalService.errorPostingChallengeModal(),
but that method is designed to show errors related to posting challenges,
so the modal’s specific message would not be appropriate for a tag-loading error.
If we want to use that approach, we should create a generic method or a more specific one.

We believe that using console.error would be appropriate, since the error would end here if we do not want to modify other parts of the code or use other options to handle it.
We can place the console.error instead of what is currently there.


```
error: (error) => {
          this.currentTags = [];
          this.tagsControl.setValue([]);
          //throw error;
          console.error('Error loading tags:', error);
        }
      });
```

ivilarop replied:
For this PR, let’s avoid throw inside the subscribe. We’re already resetting state and giving user feedback, so a minimal log is enough here.

Pablo2203 replied:
@ivilarop @FlavioKde console.error implemented!

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -93,6 +93,10 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy, OnInit
   }
   ngOnInit(): void {
+  this.tagsControl.valueChanges.subscribe(value => {
+    this.selectedTags = value || [];
```
We can simplify the state: keep currentTags for the available/tag list, and use tagsControl.value as the single source of truth for the user selection.
selectedTags is redundant (it mirrors the control via valueChanges).

**Benefits**: fewer sources of truth, less sync code, fewer edge cases.

**Action items**:

1. Remove selectedTags and read the selection from tagsControl.value wherever needed (filter, submit, UI).
2. When loading tags, filter the control’s current value against currentTags and set the filtered value once.
3. Type currentTags como Tag[] (evitar any[]).
4. Maintain consistency in errors: if tag loading fails, leave currentTags = [] and tagsControl empty/invalid.

FlavioKde replied:
Hello Ivan

What we did was adapt us to the code that it was already written. The synchronization was product of the two tag states pre existing, the only way to do not have residual tags, was the synchronization. Your comment is correct, it should be only one source of truth, but this is the way we found the code

We only fix the water leak from the pipe. If you want we can leave it as thecnical debt for future sprints

Pablo2203 replied:
@ivilarop 

Hello Ivan,

We have to add some more regarding this change you propose. The preexistent tests would fail if we did a change like that, it would be compromise this part as well. We do not know the reason and the scope of this two states inside the project.

ivilarop replied:
Thanks both for the clarification — understood 👍
The current approach works as a quick fix and keeps consistency with the existing codebase.
I fully agree we shouldn’t risk breaking pre-existing tests or UI flows in this PR.
Let’s document this as technical debt: in a future sprint we should refactor to have a single source of truth for tags (tagsControl.value) and remove the sync logic between selectedTags and currentTags.

✅ No changes required now — just make sure we mention this in the next tech-debt list or in the upcoming refactor task.

Please note that in the PR description.

Pablo2203 replied:
@FlavioKde @ivilarop Hello again! Just in case as we agreed above, I left the link with the technical debt here:
https://tree.taiga.io/project/luisvi-ita-challenges/issue/859

Cheers!

FlavioKde committed:
docs: bump version to 3.11.3

FlavioKde committed:
test: handle getTagsByLanguage errors and improve test coverage

FlavioKde committed:
Docs: bump version to 3.11.4
